### ISSUE TITLE
Document/require 'onLoad' prop for sampler type Instrument

### ISSUE THREAD

**Issue by brandongregoryscott** [unkleho/reactronica#31]
First off, I wanted to say thank you for building out this awesome library! I stumbled upon ToneJS first and figured I was going to have to build out similar abstractions for interacting with that API in React, so I'm thankful that most of the heavy lifting has been done here already.

One issue that I just spent a lot of time trying to debug was why a song with a track and sampler instrument would throw errors as soon as it attempted to play a note. I had sent through a `samples` prop with notes and file URLs in the structure shown in the docs... 

```jsx
<Instrument 
    samples={{
        "C-2": "...",
        "C#-2": "...",
    }}
    type="sampler" />
```

Any time I change `isPlaying` to `true` for the `<Song />`, it would throw this error:
```
Uncaught Error: No available buffers for note: -12
    at s.default.Sampler._findClosest (Tone.js:7)
    at s.default.Sampler.triggerAttack (Tone.js:7)
    at s.default.Sampler.triggerAttackRelease (Tone.js:7)
    at Track.tsx:95
    at Array.map (<anonymous>)
    at Track.tsx:94
    at Array.forEach (<anonymous>)
    at s.default.Sequence.callback (Track.tsx:93)
    at s.default.Sequence.s.default.Part._tick (Tone.js:7)
    at s.default.Event._tick (Tone.js:7)
    at s.default.TransportRepeatEvent.s.default.TransportEvent.invoke (Tone.js:7)
    at s.default.TransportRepeatEvent.invoke (Tone.js:7)
    at s.default.TransportEvent.invoke (Tone.js:7)
    at s.default.Timeline.<anonymous> (Tone.js:7)
    at s.default.Timeline.<anonymous> (Tone.js:7)
    at s.default.Timeline.<anonymous> (Tone.js:7)
    at Array.forEach (<anonymous>)
    at s.default.Timeline._iterate (Tone.js:7)
    at s.default.Timeline.forEachAtTime (Tone.js:7)
    at s.default.Transport._processTick (Tone.js:7)
    at s.default.Clock.<anonymous> (Tone.js:7)
    at s.default.TickSource.forEachTickBetween (Tone.js:7)
    at s.default.Clock._loop (Tone.js:7)
    at s.default.Context.s.default.Emitter.emit (Tone.js:7)
```

After some more trial and error, I realized from the network tab in dev tools that samples weren't actually being loaded. Once the `onLoad` function was provided, network requests were firing off to load the resources and then it would play as expected.

At the very least, there should be a JSDoc comment that notes this function is required when providing `type="sampler"`. It would be even better if we could use conditional types to require this property when `type` is passed through as `"sampler"`.

**Comment by brandongregoryscott** [comment]
@unkleho I'm happy to work on this, but I'm getting some errors when I try to install or build:

```
@rollup/plugin-replace: 'preventAssignment' currently defaults to false. It is recommended to set this option to `true`, as the next major version will default this option to `true`.
@rollup/plugin-replace: 'preventAssignment' currently defaults to false. It is recommended to set this option to `true`, as the next major version will default this option to `true`.
✓ Creating entry file 2.4 secs
(typescript) Error: /Users/Brandon/reactronica/src/lib/buildSequencerStep.ts(32,19): semantic error TS2349: This expression is not callable.
  Each member of the union type '(<U>(callbackfn: (value: StepNoteType, index: number, array: StepNoteType[]) => U, thisArg?: any) => U[]) | (<U>(callbackfn: (value: string, index: number, array: string[]) => U, thisArg?: any) => U[])' has signatures, but none of those signatures are compatible with each other.
 ```
 
 Is the current build meant to be passing? I can try to go through the errors and get the build passing myself, but I won't go through the effort if this is already on your radar/expected.

**Comment by unkleho** [comment]
Hi @brandongregoryscott,
Thanks for the comments!
Regarding the onLoad function, it shouldn't be required when using `sampler`. I've just done a test and able to load samples without `onLoad`. Do you have a test repo I could look at?
Perhaps the sample is long and taking some time to load?

I'm also looking into the build issue, I think I have sorted it out, silly Typescript issue on my behalf.

**Comment by unkleho** [comment]
BTW, I've just published v0.6.0, fixing the build problem and typing issues.

**Comment by brandongregoryscott** [comment]
Thanks for sorting out the TS issue and adding the `buffers` param in the callback 👍 I'll work on a codesandbox for demonstration and report back here

**Comment by brandongregoryscott** [comment]
After some fiddling in CodeSandbox, I think the issue I was running into was setting the `samples` prop after the initial render without `onLoad`. Looks like it won't try to load any new samples provided unless that prop is given too. 

See this example here: https://codesandbox.io/s/reactronica-issues-31-1cxvh 

The 'Track 3' version is ultimately what I got to, but I started with something like Track 2 where a user could pick samples and load them up on-the-fly, but it kept erroring out. I think it would still be good to document this behavior for future consumers 😅

**Comment by unkleho** [comment]
Thanks so much for that, I see the issue now. I've added `onLoad` as a condition for when new samples get added after initial render. https://github.com/unkleho/reactronica/blob/v0.6.0/src/components/Instrument.tsx#L307

I'll have a fix for it soon!

**Comment by unkleho** [comment]
Hi @brandongregoryscott 

I've just published 0.6.2 with the sample load fix and added some more efficient sample loading. Let us know if you run into any problems.

**Comment by brandongregoryscott** [comment]
Seems to be working now with sample updates & not passing in the `onLoad` function, thanks 😄 Going to close this out since the issue is resolved.
